### ISSUE TITLE
Received messages have no event names

### ISSUE THREAD

**Issue by robholmes** [ably/ably-java#366]
I've proved the message published contains an event name (using the Ably dev console), but in Android/Java `MessageListener` callback the name field is `null`. I believe this is because the string equality check is using "reference equality" `==` rather than "value equality" `String.equals()`.

Examples of this can be found:
- https://github.com/ably/ably-java/blob/master/lib/src/main/java/io/ably/lib/types/Message.java#L85
- https://github.com/ably/ably-java/blob/v1.0.1/lib/src/main/java/io/ably/lib/types/Message.java#L85
- https://github.com/ably/ably-java/blob/v1.1.0-beta.push.1/lib/src/main/java/io/ably/lib/types/Message.java#L108
- https://github.com/ably/ably-java/blob/v1.1.0-beta.push.1/lib/src/main/java/io/ably/lib/types/Message.java#L110

The fix should be to use `fieldName.equals("name")`, or `"name".equals(fieldName)` if `fieldName` could ever be `null`.

**Comment by mattheworiordan** [comment]
Thanks @robholmes for alerting us to this issue..  I will get one of the team to investigate tomorrow.

**Comment by paddybyers** [comment]
I'll have a look at this.

I don't think it's quite as simple as you say, because the `fieldName`s in the lines you highlighted are `intern()`ed (eg https://github.com/ably/ably-java/blob/v1.1.0-beta.push.1/lib/src/main/java/io/ably/lib/types/Message.java#L103). So reference equality will work when comparing with a literal string.

**Comment by robholmes** [comment]
@paddybyers I'm not sure that's true since using `fieldName == "name"` creates a new String of `"name"` and compares it with the object `fieldName`, by definition these are different objects. They may have the same value, but the only way to compare a String value is with `.equals()`.

**Comment by paddybyers** [comment]
See https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#intern--

**Comment by robholmes** [comment]
It is true that I didn't know what the `.intern()` did until now, but I believe if you really want to use `==` for a string comparison, then I think you would need to `fieldName == "name".intern()` (assuming `fieldName` has already been intern'd). At least that's my understanding, as `"name"` would not come from the unique intern pool, it's simply a new object.

**Comment by paddybyers** [comment]
Lets try to reproduce the issue and see what we find.

**Comment by robholmes** [comment]
Ok thanks Paddy, happy to help where I can. I wasn't able to go as far as writing any tests for this, but I did prove through debugging that the `if (fieldName == "name") {` was never entered, and it always fell into `else {`.

**Comment by paddybyers** [comment]
Ok, that's good information. @funkyboy here will try to reproduce it. There's something unusual about it so let's see.

**Comment by funkyboy** [comment]
@robholmes I created a [sample Android project](https://github.com/funkyboy/AblyJavaTest).
See all the detailed steps in the README file.
I could not reproduce this. 

Can you pull the repo and see if you can reproduce it in your setting? And also provide details about your setting? E.g. Java version, etc.

**Comment by robholmes** [comment]
@funkyboy ok great, I'll give this a try as soon as I can and let you know

**Comment by funkyboy** [comment]
@robholmes FYI I added you as a collaborator to [the repo](https://github.com/funkyboy/AblyJavaTest). Feel free to push your changes in a branch if needed.

**Comment by robholmes** [comment]
@funkyboy I am currently at a loss to explain it, I cannot reproduce with the sample app, even when I add all our dependencies, gradle plugins, and target/compile versions, etc. I've added the breakpoints in the same place, but it just behave like it does in our app. I'm using curl also to rule out our message publishing code.

Right now I've hit the brick wall of not knowing what I can try next. I may try cloning the ably-java repo and importing into our project so I can tweak the code and see what I can work out. I'll let you know what I find (if anything).

**Comment by funkyboy** [comment]
@robholmes Thanks for the update. Let us know if there are new developments.

**Comment by robholmes** [comment]
@funkyboy I have managed to reproduce the issue by cloning the ably-java project, building the AAR, and importing this into our project as a module.

My first step was to run the app with `master` branch as-is, and I saw the same problem of no event name due to the comparison. Second step was to implement my suggestion of `"name".equals(fieldName)`, which does correct the problem and then works as expected.

Though I'm still not sure what's special about our project that means we can't reproduce the issue in the quickstart app. There must be a fundamental difference in how the intern string pool works which means the `fieldName` and `"name"` are different objects.

Though I do believe that `.equals()` is safer, I can see no reason why `.intern()` and `==` fails in some scenarios.

I'd be happy to create a PR for consideration, but without understanding the issue I'm hesitant to proceed further.

**Comment by paddybyers** [comment]
@robholmes thanks for this.

The reason for doing `.intern()` and `==` is that a given string is being compared with multiple other literal strings, and this happens multiple times per message, and that can happen a thousand times a second or more. So I really did want `==` instead of having to do a string comparison each time.

It's not supposed to be "safer" to use `.equals()` ; the language spec says that `==` must work when comparing literal strings and/or interned strings. If this really is failing then it's a serious bug somewhere and it deserves understanding I think. That said, we're writing a library that must work, instead of fixing the JVM.

Can you share details of the specific java/Android version you're running on?

Thanks

**Comment by paddybyers** [comment]
From java 1.7 / Android 19 you can do

`switch(fieldName) {
case "name":
  ...
}
`

I'd be inclined to go that route tbh - then the compiler and VM will determine the most efficient way to implement it.

We've had 1.7 source compatibility as a requirement in the library for a while anyway.

**Comment by robholmes** [comment]
@paddybyers I completely understand the reasoning, which is another part of my hesitancy. My reasoning for stating `.equals()` is "safer", was due to it comparing the actual string contents rather than the object reference, but I agree that the `==` should work. I would like as much as you to understand this before taking any action.

Here's a snippet of our gradle config:
```gradle
    dependencies {
        classpath 'com.android.tools.build:gradle:2.3.3'
    }

    compileSdkVersion 25
    buildToolsVersion '25.0.3'

    defaultConfig {
        minSdkVersion 19
        targetSdkVersion 22
    }
```

But I don't believe this is the problem, as I copied almost all of our gradle files into the sample app @funkyboy provided, to ensure it wasn't one of our dependencies causing the issue, and I couldn't reproduce the issue.

**Comment by paddybyers** [comment]
@robholmes sure.

You're seeing this running in the emulator or device or both?

I see we're already doing `switch(string) { ... }` in `BaseMessage`. And then we have a huge `if() ... else if() ...` in `ProtocolMessage`. So this needs tidying one way or the other.

**Comment by robholmes** [comment]
@paddybyers after trying our app on the emulator (api 22 / 5.1) everything works as expected :+1: 

so it is purely related to our app running on the Samsung tablet (why doesn't this surprise me)

**Comment by paddybyers** [comment]
> so it is purely related to our app running on the Samsung tablet (why doesn't this surprise me)

:( So I think that confirms it's a JVM bug.

Can you try https://github.com/ably/ably-java/pull/367 ?

**Comment by robholmes** [comment]
@paddybyers #367 works perfectly :+1:

**Comment by robholmes** [comment]
@paddybyers that's awesome, thank you so much for your help on this one, not a simple one to work out that's for sure! :smile:

**Comment by robholmes** [comment]
as a sidenote, it was interesting to see the decompiled .class file for Message class with the changes in #367 

![image](https://user-images.githubusercontent.com/206717/36822403-c58930c6-1cef-11e8-90d6-5f7234db2786.png)

I'm no expert but this looks frustratingly less efficient now (thanks Samsung and your broken JVM implementation) :roll_eyes:

**Comment by paddybyers** [comment]
The `.hashCode()` will be roughly as expensive as the `.intern()`, and the `switch` statements are probably a decompiled `lookupswitch` bytecode construct which will generate pretty efficient native code. However it's still got the `.equals()` and I would be surprised if the VM's compiler can eliminate that.

So it's not going to be as good as just having a `.equals()` but probably not worse by a large margin.

It's disappointing that the compiler can't emit code that takes advantage of the intern mechanism.

**Comment by paddybyers** [comment]
Of course the `.equals()` does map to a natively implemented method in the VM (https://android.googlesource.com/platform/libcore/+/a7752f4d22097346dd7849b92b9f36d0a0a7a8f3/libdvm/src/main/java/java/lang/String.java?autodive=0%2F%2F#750) so there's every chance that that would short-circuit to a reference equality check if both strings happened to be interned.

**Comment by funkyboy** [comment]
Fixed by #367
